Change to expression templates for initializing operators

kalj opened the pull request:
Allt ska funka, så vi borde kunna mergea detta. Kolla exemplen för hur det används. Jag introducerade några nya klasser:

- `Stencil` -- en rad i en operator
- `BlockStencil` -- en eller flera rader i en operator
- `Operator` -- en endimensionell SBP-operator utan kännedom om nät. Innehåller en `Stencil` i det inre, och en `BlockStencil` på varje rand

Jag är inte helt nöjd med namnen på dessa -- det blir dumt med "Block" både i `BlockStencil` och i `*BlockOperator`, och kanske även med "Operator" i både `Operator` och `*BlockOperator`. Förslag på bättre namn är välkomna!

kalj committed:
need c++14 for appending arrays

kalj committed:
convert to expression-templates-based set up of stencils

kalj committed:
update examples to use new infrastructure

ylvarydin commented on include/stenseal/block_operator.h:
```diff
@@ -12,25 +12,26 @@
 
   public:
-    UpwindBlockOperator(const Geometry &g);
+    UpwindBlockOperator(const DmT dm, const Geometry &g);
```
Dm och inte Dm?

ylvarydin commented:
BlockStencil känns inte som ett helt lyckat namn eftersom vi redan har något annat som vi kallar block i strukturen. BoundaryStencil kanske vore bättre?

kalj committed:
Add documentation and clean up declarations

Document the new classes, and put all implementations away from the
declarations for readability.

kalj commented:
Lade till dokumentation.

kalj commented:
@ylvarydin Ja kanske, men eftersom vi kanske vill använda detta även för "c-stencilen" i det kompakta fallet känns boundary lite för specifikt. `StencilRows` kanske?

kalj committed:
change name of BlockStencil to StencilArray

kalj commented:
Jag och @Werpers diskuterade detta igår och kom fram till att `StencilArray` var ett okej namn tills vidare eftersom det beskriver vad det är. Pushade ny commit som fixar detta.

Jag är nöjd nu, så om nån vill mergea så är det fritt fram!

ylvarydin commented:
Jag är också nöjd :)

kalj committed:
Merge branch 'master' into expression-templates-attempt

Merge in changes from initialization of function

kalj commented:
Testar dessa merge-svartkonster. D:

kalj commented:
Nu går detta att mergea in. Gört!
